week-6/Homework/mandatory/1-alarmclock/alarmclock.js

```diff
@@ -1,7 +1,40 @@
 // Implement the missing functionality according to the task description.
 // Hint: Remember which function you need to call when it's time to alarm.
+// Global variables
+let setbtn = document.getElementById("set");
+let stopbtn = document.getElementById("stop");
+let display = document.getElementById("timeRemaining");
+let timeValue = document.getElementById("alarmSet");
+let timer;
+
+// This function will have a counter in seconds and will change the display by deducting a second everytime it's called. when the counter is 0 it will call playAlarm function
+function trigger() {
+  let counter = parseInt(display.innerText.slice(0,2) * 60) + parseInt(display.innerText.slice(3));
+  if (counter > 0 ) {
+    counter--;
+    let seconds = counter % 60;
+    let minutes =  Math.floor(counter/60);
+    display.innerText = minutes.toString().padStart(2,"0") + ":" + seconds.toString().padStart(2,"0");
+  } else if (counter === 0) {
+    playAlarm();
+  }
+}
+
+// This function will read the value inputed by user, convert it to minutes and seconds then display it and call the trigger function every second.
+function setAlarm() {
+  if (timeValue.value <60 && timeValue.value > 0){
+    display.innerText = "00:" + timeValue.value.toString().padStart(2,"0");
+  } else {
+    display.innerText = Math.floor(timeValue.value/60).toString().padStart(2,"0") + ":" + (timeValue.value % 60).toString().padStart(2,"0");
+  }
+  timer = setInterval(trigger,1000);
```

Comment:
Only one caveat, I've spotted whilst tested your page, when I click the `Set Alarm` button more than once, it doesn't reset the timer already set up :). It makes the countdown a bit glitchy as the previous timer is not terminated. Also, alarm can't be turned off because the timer that got stuck in the background keeps it turning back :)

Apart from this minor issue, your code works pretty well.

Reply:
You mean that I should clear the interval before assigning it in the code?  same way as line 34?

Reply:
I solved this with the following:
1. In setAlarm Function added one more condition as follows 
```
if(!timer){
  timer = setInterval(trigger,1000);
}
```

2.  In the eventlistener for Stop Button after clearing the interval I gave the timer a value of null
```
  clearInterval(timer); 
  timer = null;
```

Reply:
Yes, because what if you don't click the stop button before set alarm again?
Try it:
Set the time to let's say 30 sec and then click set alarm button 5 times. You will see what happens :)

Reply:
Okay I see, you fixed it now, you added and if case saying
```
if(!timer){
  timer = setInterval(trigger,1000);
}
```
Instead of shutting down the timer and creating a new one, instead, you reuse the same one if it already exists and you only update the display time.
It is a good approach too ⭐ Well done ⭐